src/query-tree/validation.ts

```diff
@@ -104,6 +104,69 @@ export class ErrorIfNotTruthyResultValidator extends QueryNode implements QueryR
     }
 }
 
+/**
+ * Validates that a TransformListQueryNode does not resolve to fewer elements
+ * than expected. This can happen when an implicit query/mutation limit is applied
+ * to the TransformListQueryNode but more elements are available in the collection.
+ *
+ * In this case a helpful error message should be returned to the consumer so that
+ * a manual query limit can be applied.
+ */
+export class NoImplicitlyTruncatedListValidator extends QueryNode implements QueryResultValidator {
+    constructor(private readonly maximumExpectedNumberOfItems: number) {
+        super();
+    }
+
+    // The following function will be translated to a string and executed within the ArangoDB server itself.
+    // Therefore the next comment is necessary to instruct our test coverage tool (https://github.com/istanbuljs/nyc)
+    // not to instrument the code with coverage instructions.
+
+    /* istanbul ignore next */
+    static getValidatorFunction() {
+        return function (validationData: any, result: any) {
+            /**
+             * An error that is thrown if a validator fails
+             */
+            class RuntimeValidationError extends Error {
+                readonly code: string | undefined;
+
+                constructor(message: string, args: { readonly code?: string } = {}) {
+                    super(message);
+                    this.name = this.constructor.name;
+                    this.code = args.code;
+                }
+            }
+
+            if (result.length > validationData.maximumExpectedNumberOfItems) {
+                throw new RuntimeValidationError(
+                    `Collection is truncated by default to ${validationData.maximumExpectedNumberOfItems} elements but contains more elements than this limit. Specify a limit manually to retrieve all elements of the collection.`,
```

Comment:
I would not use the term "truncated" here because for the user, nothing is truncated - an error is thrown. Implicit truncation is exactly what we want to avoid doing.

There's also a second reason you might want to specify a limit: if you _want_ a truncated result. So I would include that in the error message.

I think phrasing the error is easier when we know the kind of operation we do

'Query would return more than {limit} objects. Specify "first" to increase the limit or truncate the result.'

'Mutation would return more than {limit} objects. Specify "first" to increase the limit or truncate the result.'

Reply:
> Wouldn't the phrasing for the mutation case better contain something along the lines of "Mutation would modify more than ${limit} objects (...)" since the modification is already the problem?

good question. Currently, there is no way to modify objects without returning every one of them. If there were such a way, we might be able to use higher limits there because just modifying is less of a problem than serializing the result and passing it everywhere.

As we don't have this API yet, maybe "would modify" is indeed simpler to understand in this context. Or just "affect"? "Modify" sounds a bit like this is the number of objects that are actually modified (with changes). It also would work better for delete.